evap/results/views.py

```diff
@@ -135,13 +136,26 @@ def index(request):
         additional_evaluations = get_evaluations_with_course_result_attributes(additional_evaluations)
         evaluations += additional_evaluations
 
-    evaluations.sort(key=lambda evaluation: (evaluation.course.semester.pk, evaluation.full_name))  # evaluations must be sorted for regrouping them in the template
+    # put evaluations into a dict that maps from course to a list of evaluations.
+    # this dict is sorted by course.pk (important for the zip below)
+    # (this relies on python 3.7's guarantee that the insertion order of the dict is preserved)
+    evaluations.sort(key=lambda evaluation: evaluation.course.pk)
 
-    evaluation_pks = [evaluation.pk for evaluation in evaluations]
-    degrees = Degree.objects.filter(courses__evaluations__pk__in=evaluation_pks).distinct()
-    course_types = CourseType.objects.filter(courses__evaluations__pk__in=evaluation_pks).distinct()
+    courses = defaultdict(list)
+    for evaluation in evaluations:
+        courses[evaluation.course].append(evaluation)
+
+    course_pks = list([course.pk for course in courses.keys()])
+
+    # annotate each course in courses with num_evaluations
+    annotated_courses = Course.objects.filter(pk__in=course_pks).annotate(num_evaluations=Count('evaluations')).order_by('pk').defer()
+    for course, annotated_course in zip(courses.keys(), annotated_courses):
+        course.num_evaluations = annotated_course.num_evaluations
+
+    degrees = Degree.objects.filter(courses__pk__in=course_pks).distinct()
+    course_types = CourseType.objects.filter(courses__pk__in=course_pks).distinct()
     template_data = dict(
-        evaluations=evaluations,
+        courses=courses.items(),
```

Comment:
Waybe name the template variable `course_evaluation_tuples` or something? I was confused in the template that you didn't access it with `.items` and only saw later that you do it here.

A rather weak opinion of mine, though, so I'm also fine with leaving it like this.

Reply:
I called this courses_and_evaluations, but it's still the same in template and view. actually I wanted to do the `.items()` in the template but that does not work, seems like i still haven't understood the rules how method calls work there...

Reply:
Hmm, strange. A bit of debugging indicates that using `.items` on a `defaultdict` in a template breaks (returns an empty list), if you convert it to a normal dict first, it works fine.

I guess I'm fine with the current code.